zenoh/src/net/routing/hat/client/interests.rs

```diff
@@ -0,0 +1,173 @@
+//
+// Copyright (c) 2023 ZettaScale Technology
+//
+// This program and the accompanying materials are made available under the
+// terms of the Eclipse Public License 2.0 which is available at
+// http://www.eclipse.org/legal/epl-2.0, or the Apache License, Version 2.0
+// which is available at https://www.apache.org/licenses/LICENSE-2.0.
+//
+// SPDX-License-Identifier: EPL-2.0 OR Apache-2.0
+//
+// Contributors:
+//   ZettaScale Zenoh Team, <zenoh@zettascale.tech>
+//
+use std::sync::{atomic::Ordering, Arc};
+
+use zenoh_protocol::{
+    core::WhatAmI,
+    network::{
+        declare::ext,
+        interest::{InterestId, InterestMode, InterestOptions},
+        Declare, DeclareBody, DeclareFinal, Interest,
+    },
+};
+use zenoh_sync::get_mut_unchecked;
+
+use super::{face_hat, face_hat_mut, HatCode, HatFace};
+use crate::net::routing::{
+    dispatcher::{
+        face::{FaceState, InterestState},
+        resource::Resource,
+        tables::{Tables, TablesLock},
+    },
+    hat::{CurrentFutureTrait, HatInterestTrait},
+    RoutingContext,
+};
+
+pub(super) fn interests_new_face(tables: &mut Tables, face: &mut Arc<FaceState>) {
+    for mut src_face in tables
+        .faces
+        .values()
+        .cloned()
+        .collect::<Vec<Arc<FaceState>>>()
+    {
+        if face.whatami != WhatAmI::Client {
```

Comment:
The filter could be applied to the iterator above. In this way you avoid some unnecessary cloning when the `face.whatami != WhatAmI::Client`.

Reply:
Well, even better, as this check only applies on the function parameter we can do the check once before iterating. Fixed.